src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+            {
+                return curlioerr.CURLIOE_UNKNOWNCMD;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                CheckDisposed();
+                state.CancellationToken.ThrowIfCancellationRequested();
+
+                Stream contentStream = state.RequestContentStream;
+                if (!contentStream.CanSeek)
+                {
+                    return curlioerr.CURLIOE_FAILRESTART;
+                }
+                contentStream.Seek(0, SeekOrigin.Begin);
+                return curlioerr.CURLIOE_OK;
+            }
+            catch (Exception ex)
+            {
+                state.SavedException = ex;
+#if DEBUG
+                if (Logging.On) Logging.Exception(Logging.Http, this, "CurlSendIoCtlCallback exception for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + ": ", ex);
+#endif
+                return curlioerr.CURLIOE_FAILRESTART; // Force curl_easy_perform to fail
+            }
+        }
+
+        private ulong ExecuteCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context,
+            Func<IntPtr, ulong, RequestCompletionSource, ulong> callback)
+        {
+            size *= nitems;
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                CheckDisposed();
+#if DEBUG
+                if (Logging.On)
+                {
+                    Logging.PrintInfo(Logging.Http, this, "Executecallback for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + " with " + ((size > 0) ? (size + " bytes") : " no data"));
+                }
+#endif
+                return (size == 0) ? 0 : callback(buffer, size, state);
+            }
+            catch (Exception ex)
+            {
+                state.SavedException = ex;
+#if DEBUG
+                if (Logging.On) Logging.Exception(Logging.Http, this, "Executecallback exception for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + ": ", ex);
+#endif
+                return (size > 0) ? size - 1 : 1; // Force curl_easy_perform to fail
+            }
+        }
+
+        private bool TryParseStatusLine(HttpResponseMessage response, string responseHeader)
+        {
+
+            if (!responseHeader.StartsWith("HTTP/", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+            var tokens = responseHeader.Split(new char[] { ' ' });
+            if (tokens.Length > 1)
+            {
+                response.StatusCode = (HttpStatusCode)int.Parse(tokens[1]);
+                if (tokens.Length > 2)
+                {
+                    response.ReasonPhrase = String.Join(" ", tokens, 2, tokens.Length - 2);
+                }
```

Comment:
It seems like the string.Split is being used to make the parsing easier, but for a library like this, it'd be great to avoid as many allocations as possible, and in this case that appears to mean the results array plus the individual strings filling the array (the only necessary one being the first, because we don't have an int.Parse that works with a string and a range, unfortunately, though if it became impactful we could add here until one is available in the core libs). Seems like all of this logic could be replaced with something like (pseudo-code):
```C#
int firstSpace = responseHeader.IndexOf(' ');
if (firstSpace > 0)
{
    response.StatusCode = (HttpStatusCode)int.Parse(responseHeader.Substring(0, firstSpace);
    if (firstSpace < responseHeader.Length)
    {
        response.ReasonPhrase = responseHeader.Substring(firstSpace+1);
    }
}
```
or something like that.